### ISSUE TITLE
Balancer advices invalid PG moves

### ISSUE THREAD

**Issue by Dragonn** [TheJJ/ceph-balancer#5]
Sometimes balancer gives output containing upmap commands which are not valid according to active CRUSH rule.

I have 5 node cluster with replicated rules only with rack as failure domain. Many of generated commands are completely OK and leads to better balancing, but sometimes command (or its part) is not valid and Ceph doesn't insert it into configuration (silently, which is a bit confusing).

Cluster has following topology:
```
-> ceph osd tree
ID   CLASS  WEIGHT    TYPE NAME                      STATUS  REWEIGHT  PRI-AFF
 -1         26.17169  root default                                            
-13         26.17169      datacenter dc1                                
-14         26.17169          room room1                                       
-15          5.23602              rack rack2                                  
 -3          5.23602                  host ceph1                           
  0    ssd   0.87209                      osd.0          up   1.00000  1.00000
  2    ssd   0.87279                      osd.2          up   1.00000  1.00000
  3    ssd   0.87279                      osd.3          up   1.00000  1.00000
  4    ssd   0.87279                      osd.4          up   1.00000  1.00000
 28    ssd   0.87279                      osd.28         up   1.00000  1.00000
 29    ssd   0.87279                      osd.29         up   1.00000  1.00000
-16          5.23392              rack rack3                                  
 -5          5.23392                  host ceph2                           
  5    ssd   0.87209                      osd.5          up   1.00000  1.00000
  6    ssd   0.87209                      osd.6          up   1.00000  1.00000
  7    ssd   0.87209                      osd.7          up   1.00000  1.00000
  8    ssd   0.87209                      osd.8          up   1.00000  1.00000
 26    ssd   0.87279                      osd.26         up   1.00000  1.00000
 27    ssd   0.87279                      osd.27         up   1.00000  1.00000
-17          5.23392              rack rack5                                  
 -7          5.23392                  host ceph3                           
  1    ssd   0.87209                      osd.1          up   1.00000  1.00000
 11    ssd   0.87209                      osd.11         up   1.00000  1.00000
 14    ssd   0.87209                      osd.14         up   1.00000  1.00000
 17    ssd   0.87209                      osd.17         up   1.00000  1.00000
 24    ssd   0.87279                      osd.24         up   1.00000  1.00000
 25    ssd   0.87279                      osd.25         up   1.00000  1.00000
-18          5.23392              rack rack6                                  
 -9          5.23392                  host ceph4                           
  9    ssd   0.87209                      osd.9          up   1.00000  1.00000
 12    ssd   0.87209                      osd.12         up   1.00000  1.00000
 15    ssd   0.87209                      osd.15         up   1.00000  1.00000
 18    ssd   0.87209                      osd.18         up   1.00000  1.00000
 22    ssd   0.87279                      osd.22         up   1.00000  1.00000
 23    ssd   0.87279                      osd.23         up   1.00000  1.00000
-25          5.23392              rack rack7                                  
-11          5.23392                  host ceph5                           
 10    ssd   0.87209                      osd.10         up   1.00000  1.00000
 13    ssd   0.87209                      osd.13         up   1.00000  1.00000
 16    ssd   0.87209                      osd.16         up   1.00000  1.00000
 19    ssd   0.87209                      osd.19         up   1.00000  1.00000
 20    ssd   0.87279                      osd.20         up   1.00000  1.00000
 21    ssd   0.87279                      osd.21         up   1.00000  1.00000
```

```
-> ./placementoptimizer.py -v balance --max-pg-moves 12 | tee 2021-12-02_balance-upmaps_2
[2021-12-02 16:51:16,001] running pg balancer
[2021-12-02 16:51:16,008] current OSD fill rate per crushclasses:
[2021-12-02 16:51:16,008]   ssd: average=53.85%, median=52.51%, without_placement_constraints=51.39%
[2021-12-02 16:51:16,009] cluster variance for crushclasses:
[2021-12-02 16:51:16,009]   ssd: 12.017
[2021-12-02 16:51:16,009] min osd.22 48.749%
[2021-12-02 16:51:16,009] max osd.11 59.951%
[2021-12-02 16:51:16,013]   SAVE move 6.1f osd.11 => osd.8 (size=8.5G)
[2021-12-02 16:51:16,013]     => variance new=11.36060690389384 < 12.016851844400833=old
[2021-12-02 16:51:16,013]     new min osd.22 48.749%
[2021-12-02 16:51:16,013]         max osd.25 59.655%
[2021-12-02 16:51:16,013]     new cluster variance:
[2021-12-02 16:51:16,013]       ssd: 11.361
[2021-12-02 16:51:16,016]   SAVE move 8.1e osd.25 => osd.22 (size=12.4G)
[2021-12-02 16:51:16,016]     => variance new=10.447990002211988 < 11.36060690389384=old
[2021-12-02 16:51:16,016]     new min osd.8 49.952%
[2021-12-02 16:51:16,016]         max osd.4 59.171%
[2021-12-02 16:51:16,016]     new cluster variance:
[2021-12-02 16:51:16,017]       ssd: 10.448
[2021-12-02 16:51:16,020]   SAVE move 8.64 osd.4 => osd.8 (size=12.6G)
[2021-12-02 16:51:16,020]     => variance new=9.687735955065486 < 10.447990002211988=old
[2021-12-02 16:51:16,020]     new min osd.18 50.000%
[2021-12-02 16:51:16,020]         max osd.12 59.057%
[2021-12-02 16:51:16,020]     new cluster variance:
[2021-12-02 16:51:16,020]       ssd: 9.688
[2021-12-02 16:51:16,024]   SAVE move 8.ac osd.12 => osd.21 (size=12.4G)
[2021-12-02 16:51:16,024]     => variance new=8.957778334870069 < 9.687735955065486=old
[2021-12-02 16:51:16,024]     new min osd.18 50.000%
[2021-12-02 16:51:16,024]         max osd.11 58.999%
[2021-12-02 16:51:16,024]     new cluster variance:
[2021-12-02 16:51:16,024]       ssd: 8.958
[2021-12-02 16:51:16,027]   SAVE move 6.11 osd.11 => osd.18 (size=8.5G)
[2021-12-02 16:51:16,027]     => variance new=8.431644281707726 < 8.957778334870069=old
[2021-12-02 16:51:16,028]     new min osd.22 50.140%
[2021-12-02 16:51:16,028]         max osd.25 58.265%
[2021-12-02 16:51:16,028]     new cluster variance:
[2021-12-02 16:51:16,028]       ssd: 8.432
[2021-12-02 16:51:16,031]   SAVE move 8.d9 osd.25 => osd.22 (size=12.4G)
[2021-12-02 16:51:16,031]     => variance new=7.787358409360783 < 8.431644281707726=old
[2021-12-02 16:51:16,031]     new min osd.1 50.210%
[2021-12-02 16:51:16,031]         max osd.11 58.052%
[2021-12-02 16:51:16,031]     new cluster variance:
[2021-12-02 16:51:16,031]       ssd: 7.787
[2021-12-02 16:51:16,035]   SAVE move 6.e osd.11 => osd.1 (size=8.4G)
[2021-12-02 16:51:16,035]     => variance new=7.337994734535253 < 7.787358409360783=old
[2021-12-02 16:51:16,035]     new min osd.9 50.500%
[2021-12-02 16:51:16,035]         max osd.6 57.947%
[2021-12-02 16:51:16,035]     new cluster variance:
[2021-12-02 16:51:16,036]       ssd: 7.338
[2021-12-02 16:51:16,039]   SAVE move 8.f1 osd.6 => osd.9 (size=13.0G)
[2021-12-02 16:51:16,039]     => variance new=6.735266750261725 < 7.337994734535253=old
[2021-12-02 16:51:16,039]     new min osd.18 50.948%
[2021-12-02 16:51:16,039]         max osd.4 57.760%
[2021-12-02 16:51:16,039]     new cluster variance:
[2021-12-02 16:51:16,039]       ssd: 6.735
[2021-12-02 16:51:16,043]   SAVE move 8.1a osd.4 => osd.29 (size=12.5G)
[2021-12-02 16:51:16,043]     => variance new=6.215466556962361 < 6.735266750261725=old
[2021-12-02 16:51:16,043]     new min osd.18 50.948%
[2021-12-02 16:51:16,043]         max osd.12 57.665%
[2021-12-02 16:51:16,043]     new cluster variance:
[2021-12-02 16:51:16,043]       ssd: 6.215
[2021-12-02 16:51:16,046]   SAVE move 8.2c osd.12 => osd.14 (size=12.4G)
[2021-12-02 16:51:16,046]     => variance new=5.712574199748361 < 6.215466556962361=old
[2021-12-02 16:51:16,046]     new min osd.18 50.948%
[2021-12-02 16:51:16,047]         max osd.16 57.660%
[2021-12-02 16:51:16,047]     new cluster variance:
[2021-12-02 16:51:16,047]       ssd: 5.713
[2021-12-02 16:51:16,050]   SAVE move 6.1f osd.16 => osd.18 (size=8.5G)
[2021-12-02 16:51:16,050]     => variance new=5.334505158854212 < 5.712574199748361=old
[2021-12-02 16:51:16,050]     new min osd.26 51.127%
[2021-12-02 16:51:16,050]         max osd.11 57.107%
[2021-12-02 16:51:16,050]     new cluster variance:
[2021-12-02 16:51:16,050]       ssd: 5.335
[2021-12-02 16:51:16,054]   SAVE move 6.8 osd.11 => osd.26 (size=8.4G)
[2021-12-02 16:51:16,054]     => variance new=5.006761816792269 < 5.334505158854212=old
[2021-12-02 16:51:16,054]     new min osd.1 51.155%
[2021-12-02 16:51:16,054]         max osd.23 57.032%
[2021-12-02 16:51:16,054]     new cluster variance:
[2021-12-02 16:51:16,054]       ssd: 5.007
[2021-12-02 16:51:16,054] enough remaps found
[2021-12-02 16:51:16,054] --------------------------------------------------------------------------------
[2021-12-02 16:51:16,054] generated 12 remaps.
[2021-12-02 16:51:16,054] total movement size: 130.1G.
[2021-12-02 16:51:16,054] --------------------------------------------------------------------------------
[2021-12-02 16:51:16,054] old cluster variance per crushclass:
[2021-12-02 16:51:16,055]   ssd: 12.017
[2021-12-02 16:51:16,055] old min osd.22 48.749%
[2021-12-02 16:51:16,055] old max osd.11 59.951%
[2021-12-02 16:51:16,055] --------------------------------------------------------------------------------
[2021-12-02 16:51:16,055] new min osd.1 51.155%
[2021-12-02 16:51:16,055] new max osd.23 57.032%
[2021-12-02 16:51:16,055] new cluster variance:
[2021-12-02 16:51:16,055]   ssd: 5.007
[2021-12-02 16:51:16,055] --------------------------------------------------------------------------------
ceph osd pg-upmap-items 6.1f 11 8 16 18
ceph osd pg-upmap-items 8.1e 25 22
ceph osd pg-upmap-items 8.64 4 8
ceph osd pg-upmap-items 8.ac 4 19 12 21
ceph osd pg-upmap-items 6.11 11 18
ceph osd pg-upmap-items 8.d9 25 22
ceph osd pg-upmap-items 6.e 11 1
ceph osd pg-upmap-items 8.f1 6 9
ceph osd pg-upmap-items 8.1a 11 9 4 29
ceph osd pg-upmap-items 8.2c 25 26 12 14
ceph osd pg-upmap-items 6.8 11 26
```

But e.g. second move of PG `6.1F` would violate CRUSH rule because it would colocate 2nd and 3rd replica to same host `ceph4`:
```
-> ceph pg dump | grep -F 6.1f
dumped all
6.1f        2262                   0         0          0        0   9118765056            0           0  2918      2918  active+clean  2021-12-02T01:29:46.271762+0000    1043596'3570596   1043603:54558341  [11,12,16]          11  [11,12,16]              11    1034824'3565764  2021-12-01T16:09:13.967493+0000    1023825'3552322  2021-11-30T10:06:06.894016+0000              0
```

And also Ceph will (silently) refuse to proceed command when I try to run them:
```
-> ceph osd dump | grep -F 6.1f

-> ceph osd pg-upmap-items 6.1f 11 8 16 18
set 6.1f pg_upmap_items mapping to [11->8,16->18]

-> ceph osd dump | grep -F 6.1f
```

But I can do first relocation just fine:
```
-> ceph osd pg-upmap-items 6.1f 11 8
set 6.1f pg_upmap_items mapping to [11->8]

-> ceph osd dump | grep -F 6.1f
pg_upmap_items 6.1f [11,8]
pg_temp 6.1f [11,12,16]
```

Please let me know if I can be somewhat helpful with solving this issue.

**Comment by TheJJ** [comment]
That's bad. And it means my crush validation is somehow wrong. I guess that's also why #4 occurs, and the sanity check fires.
It's really hard debugging without a working cluster currently, I should add a feature where it collects all the json data so you may send it to me and one can test/fix it completely offline.

Also here, for now, it would be helpful if you ran with maximum logging, and send me the output of the whole decision process of one move that'll be invalid, and one of a move that is valid.

**Comment by Dragonn** [comment]
Okay, sorry I should read both your responses before replying :-).

Currently we use cluster in a way that 1 rack = 1 host, but we are currently preparing to extend cluster which will result in multiple hosts within single rack so I cannot really change CRUSH topology right now.

However I think I can send you all necessary JSON outputs which are collected in the beginning of script. I just probably prefer not to attach them publicly to the issue directly. Can we find some more private channel to deliver them? E.g. email or PM somewhere?

**Comment by TheJJ** [comment]
I've now added a feature to export the state bundle :)
Please create it with `./placementoptimizer.py gather outputfile.xz` (you can reach me in matrix or via mail - `jj -at- sft lol` :)

**Comment by dimm0** [comment]
Same here, almost all moves are not allowed in a cluster with datacenter failure domain

**Comment by TheJJ** [comment]
Are running the latest version?

Can you send me your state dump, like i described previously, please? Also, for a given cluster state dump, please send me generated moves that ceph does not accept, but which the balancer produced. That would mean i can tinker with the code and try to figure out the problem.

**Comment by dimm0** [comment]
Yes, the most recent.

I had to do a couple changes, since json is giving a wrong type for replicated pools (I'm using rook, it might've done something weird in config):

```
diff --git a/placementoptimizer.py b/placementoptimizer.py
index 91a37ae..25a64e6 100755
--- a/placementoptimizer.py
+++ b/placementoptimizer.py
@@ -320,17 +320,18 @@ for pool in CLUSTER_STATE["df_dump"]["pools"]:
 
 
 for pool in CLUSTER_STATE["pool_dump"]:
+
     id = pool["pool_id"]
     ec_profile = pool["erasure_code_profile"]
 
     pg_shard_size_avg = pools[id]["stored"] / pools[id]["pg_num"]
 
-    if ec_profile:
+    if ec_profile and ec_profiles.get(ec_profile):
         pg_shard_size_avg /= ec_profiles[ec_profile]["data_chunks"]
 
     pools[id].update({
         "erasure_code_profile": ec_profile,
-        "repl_type": "ec" if ec_profile else "repl",
+        "repl_type": "ec" if ec_profile and ec_profiles.get(ec_profile) else "repl",
         "pg_shard_size_avg": pg_shard_size_avg
     })
 
@@ -678,7 +679,7 @@ def get_pg_shardsize(pgid):
     shard_size += pg_stats['num_omap_bytes']
 
     ec_profile = pg_ec_profile(pgid)
-    if ec_profile:
+    if ec_profile and ec_profiles.get(ec_profile):
         shard_size /= ec_profiles[ec_profile]["data_chunks"]
         # omap is not supported on EC pools (yet)
         # when it is, check how the omap data is spread (replica or also ec?)
```

```
[root@k8s-haosu-14 ~]# ./placementoptimizer.py -v balance --max-pg-moves 10
[2022-03-15 17:13:26,040] gathering cluster state via ceph api...
[2022-03-15 17:13:31,988] running pg balancer
[2022-03-15 17:13:32,023] current OSD fill rate per crushclasses:
[2022-03-15 17:13:32,024]   hdd: average=51.24%, median=50.80%, without_placement_constraints=58.60%
[2022-03-15 17:13:32,024] cluster variance for crushclasses:
[2022-03-15 17:13:32,024]   hdd: 54.884
[2022-03-15 17:13:32,025] min osd.26 37.306%
[2022-03-15 17:13:32,025] max osd.34 68.550%
[2022-03-15 17:13:32,033]   SAVE move 1.345 osd.34 => osd.26
[2022-03-15 17:13:32,033]     props: size=81.0G remapped=False upmaps=0
[2022-03-15 17:13:32,033]     => variance new=54.63124092496721 < 54.884087241088345=old
[2022-03-15 17:13:32,034]     new min osd.19 37.488%
[2022-03-15 17:13:32,034]         max osd.29 68.065%
[2022-03-15 17:13:32,034]     new cluster variance:
[2022-03-15 17:13:32,034]       hdd: 54.631
[2022-03-15 17:13:32,043]   SAVE move 1.21b osd.29 => osd.19
[2022-03-15 17:13:32,043]     props: size=81.0G remapped=False upmaps=0
[2022-03-15 17:13:32,043]     => variance new=54.384077250915475 < 54.63124092496721=old
[2022-03-15 17:13:32,043]     new min osd.170 38.010%
[2022-03-15 17:13:32,043]         max osd.34 67.683%
[2022-03-15 17:13:32,043]     new cluster variance:
[2022-03-15 17:13:32,043]       hdd: 54.384
[2022-03-15 17:13:32,052]   SAVE move 1.10c osd.34 => osd.170
[2022-03-15 17:13:32,052]     props: size=80.9G remapped=False upmaps=0
[2022-03-15 17:13:32,053]     => variance new=54.161715943167025 < 54.384077250915475=old
[2022-03-15 17:13:32,053]     new min osd.26 38.170%
[2022-03-15 17:13:32,053]         max osd.29 67.199%
[2022-03-15 17:13:32,053]     new cluster variance:
[2022-03-15 17:13:32,053]       hdd: 54.162
[2022-03-15 17:13:32,062]   SAVE move 1.75 osd.29 => osd.26
[2022-03-15 17:13:32,062]     props: size=80.2G remapped=False upmaps=0
[2022-03-15 17:13:32,062]     => variance new=53.9295038278412 < 54.161715943167025=old
[2022-03-15 17:13:32,062]     new min osd.133 38.245%
[2022-03-15 17:13:32,062]         max osd.34 66.817%
[2022-03-15 17:13:32,062]     new cluster variance:
[2022-03-15 17:13:32,062]       hdd: 53.930
[2022-03-15 17:13:32,071]   SAVE move 1.1a0 osd.34 => osd.133
[2022-03-15 17:13:32,071]     props: size=80.4G remapped=False upmaps=0
[2022-03-15 17:13:32,071]     => variance new=53.71749631647796 < 53.9295038278412=old
[2022-03-15 17:13:32,071]     new min osd.122 38.325%
[2022-03-15 17:13:32,071]         max osd.29 66.341%
[2022-03-15 17:13:32,071]     new cluster variance:
[2022-03-15 17:13:32,071]       hdd: 53.717
[2022-03-15 17:13:32,081]   SAVE move 1.38f osd.29 => osd.122
[2022-03-15 17:13:32,081]     props: size=80.2G remapped=False upmaps=0
[2022-03-15 17:13:32,082]     => variance new=53.510389448720424 < 53.71749631647796=old
[2022-03-15 17:13:32,082]     new min osd.19 38.353%
[2022-03-15 17:13:32,082]         max osd.34 65.956%
[2022-03-15 17:13:32,082]     new cluster variance:
[2022-03-15 17:13:32,082]       hdd: 53.510
[2022-03-15 17:13:32,090]   SAVE move 1.fd osd.34 => osd.19
[2022-03-15 17:13:32,091]     props: size=80.0G remapped=False upmaps=0
[2022-03-15 17:13:32,091]     => variance new=53.290570485934 < 53.510389448720424=old
[2022-03-15 17:13:32,091]     new min osd.70 38.369%
[2022-03-15 17:13:32,091]         max osd.29 65.484%
[2022-03-15 17:13:32,091]     new cluster variance:
[2022-03-15 17:13:32,091]       hdd: 53.291
[2022-03-15 17:13:32,100]   SAVE move 1.1e8 osd.29 => osd.70
[2022-03-15 17:13:32,100]     props: size=80.2G remapped=False upmaps=0
[2022-03-15 17:13:32,100]     => variance new=53.09083806042146 < 53.290570485934=old
[2022-03-15 17:13:32,100]     new min osd.130 38.645%
[2022-03-15 17:13:32,100]         max osd.53 65.476%
[2022-03-15 17:13:32,100]     new cluster variance:
[2022-03-15 17:13:32,100]       hdd: 53.091
[2022-03-15 17:13:32,109]   SAVE move 1.11e osd.53 => osd.130
[2022-03-15 17:13:32,109]     props: size=80.5G remapped=False upmaps=0
[2022-03-15 17:13:32,109]     => variance new=52.892366525271036 < 53.09083806042146=old
[2022-03-15 17:13:32,109]     new min osd.170 38.730%
[2022-03-15 17:13:32,109]         max osd.34 65.101%
[2022-03-15 17:13:32,109]     new cluster variance:
[2022-03-15 17:13:32,109]       hdd: 52.892
[2022-03-15 17:13:32,118]   SAVE move 28.6de osd.34 => osd.170
[2022-03-15 17:13:32,118]     props: size=48.1G remapped=False upmaps=0
[2022-03-15 17:13:32,118]     => variance new=52.77434199993548 < 52.89236652527104=old
[2022-03-15 17:13:32,118]     new min osd.140 38.759%
[2022-03-15 17:13:32,118]         max osd.66 65.037%
[2022-03-15 17:13:32,118]     new cluster variance:
[2022-03-15 17:13:32,118]       hdd: 52.774
[2022-03-15 17:13:32,118] enough remaps found
[2022-03-15 17:13:32,118] --------------------------------------------------------------------------------
[2022-03-15 17:13:32,118] generated 10 remaps.
[2022-03-15 17:13:32,118] total movement size: 772.4G.
[2022-03-15 17:13:32,118] --------------------------------------------------------------------------------
[2022-03-15 17:13:32,119] old cluster variance per crushclass:
[2022-03-15 17:13:32,119]   hdd: 54.884
[2022-03-15 17:13:32,119] old min osd.26 37.306%
[2022-03-15 17:13:32,119] old max osd.34 68.550%
[2022-03-15 17:13:32,119] --------------------------------------------------------------------------------
[2022-03-15 17:13:32,119] new min osd.140 38.759%
[2022-03-15 17:13:32,119] new max osd.66 65.037%
[2022-03-15 17:13:32,119] new cluster variance:
[2022-03-15 17:13:32,119]   hdd: 52.774
[2022-03-15 17:13:32,119] --------------------------------------------------------------------------------
ceph osd pg-upmap-items 1.345 34 26
ceph osd pg-upmap-items 1.21b 29 19
ceph osd pg-upmap-items 1.10c 34 170
ceph osd pg-upmap-items 1.75 29 26
ceph osd pg-upmap-items 1.1a0 34 133
ceph osd pg-upmap-items 1.38f 29 122
ceph osd pg-upmap-items 1.fd 34 19
ceph osd pg-upmap-items 1.1e8 29 70
ceph osd pg-upmap-items 1.11e 53 130
ceph osd pg-upmap-items 28.6de 34 170
```

```
1.345      20898                   0         0          0        0   86954708992          296          27  3386      3386                        active+clean  2022-03-14T05:26:58.567934+0000   1908585'13965079   1908585:39935681            [99,34,104]          99            [99,34,104]
```

I have the datacenter failure domain

1.345 34->26 will move it from `datacenter ucsb` to `datacenter ucsd` which is also the datacenter of osd.99 (in fact the same node even)

**Comment by dimm0** [comment]
I'm [@dimm:matrix.nrp-nautilus.io](https://matrix.to/#/@dimm:matrix.nrp-nautilus.io) on matrix

**Comment by dimm0** [comment]
Oh, I just read in another ticket that you only support hosts for now. I guess that's the issue then

**Comment by TheJJ** [comment]
It seems that it is the case, but I do want to fix the bugs so we can use it to balance arbitrary crush trees :) I've contacted you on matrix!

**Comment by dimm0** [comment]
I've got the matrix invite and messaged you, and see no other messages from you. Not sure if you're off somewhere or my matrix server is not happy

**Comment by TheJJ** [comment]
Wohoo I've fixed this in d92c2fcbc42794e8e55fc7b1a016df7929d0f9d3 ! Thanks for your perfect reproduction cluster state file @dimm0 :smile:
